### PR TITLE
refactor: CLI subcommands and argv

### PR DESCRIPTION
**Pull request by bartlomieju**
Main goal of this refactor is to eliminate `main_module: Option<String>` fields and replace them with `main_module: String`. As a side effect it allows to eliminate lot of code in tests and cleanup `flags.argv` to not contain "deno" everytime...

### CODE REVIEW

File: cli/ops/runtime.rs [hunk 1]
```diff
@@ -28,16 +29,13 @@ fn op_start(
...
-    vec![]
-  };
+
   Ok(JsonOp::Sync(json!({
     "cwd": deno_fs::normalize_path(&env::current_dir().unwrap()),
     "pid": std::process::id(),
-    "argv": script_args,
-    "mainModule": gs.main_module.as_ref().map(|x| x.to_string()),
+    "args": gs.flags.argv.clone(),
+    "repl": gs.flags.subcommand == DenoSubcommand::Repl,
```
Comment (bartlomieju) [inline cli/ops/runtime.rs]: `repl` is very good candidate for a separate entry point - just like we did with `worker` - that should help working on it in the future.

File: cli/js/runtime_main.ts [hunk 2]
```diff
@@ -72,14 +72,13 @@ export function bootstrapMainRuntime(): void {
...
+    args.push(s.args[i]);
   }
   log("args", args);
   Object.freeze(args);
 
-  // TODO(bartlomieju): rename to s.repl
-  if (!s.mainModule) {
+  if (s.repl) {
     replLoop();
   }
```
Comment (bartlomieju) [inline cli/js/runtime_main.ts]: `REPL` is a very good candidate for a separate entry point - just like we did with worker - that should help working on it in the future.

Comment (bartlomieju) [reply cli/js/runtime_main.ts]: Ref #3760

File: cli/flags.rs [hunk 3]
```diff
@@ -355,10 +368,10 @@ fn completions_parse(flags: &mut DenoFlags, matches: &clap::ArgMatches) {
...
     &mut buf,
   );
-  // TODO(ry) This flags module should only be for parsing flags, not actually
-  // acting upon the flags. Although this print is innocent, it breaks the
-  // model. The print should be moved out.
-  print!("{}", std::str::from_utf8(&buf).unwrap());
+
+  flags.subcommand = DenoSubcommand::Completions {
+    buf: buf.into_boxed_slice(),
+  };
```
Comment (ry) [inline cli/flags.rs]: Cool

**Pull request merged**